### PR TITLE
Replace CUDA API wrapper memory operations with native CUDA calls

### PR DESCRIPTION
**Pull request by waredjeb**
#### PR description 
This PR is part of #386:
 -  replace `cuda::memory::copy()` with `cudaMemcpy()`, c`uda::memory::async::copy()` with `cudaMemcpyAsync()`
  - replace `cuda::memory::zero()` and `cuda::memory::set()` with `cudaMemset()`
  - replace `cuda::memory::async::zero()` and `cuda::memory::async::set()` with `cudaMemsetAsync()`

#### PR validation 
unit tests run

### CODE REVIEW

File: CUDADataFormats/BeamSpot/src/BeamSpotCUDA.cc [hunk 1]
```diff
@@ -4,5 +4,5 @@
 
 BeamSpotCUDA::BeamSpotCUDA(Data const* data_h, cuda::stream_t<>& stream) {
   data_d_ = cudautils::make_device_unique<Data>(stream);
-  cuda::memory::async::copy(data_d_.get(), data_h, sizeof(Data), stream.id());
+  cudaMemcpyAsync(data_d_.get(), data_h, sizeof(Data), cudaMemcpyHostToDevice, stream.id());
```
Comment (makortel) [inline CUDADataFormats/BeamSpot/src/BeamSpotCUDA.cc]: We've typically used `cudaMemcpyDefault` elsewhere (but I'm not against of denoting the direction explicitly).

**Comment by makortel** [comment]
There are conflicts so this PR needs to be rebased (OTOH this PR conflicts also with #389, so it could be less work to wait until that one gets merged, I let @fwyzard to comment whether #389 could get in soon).

File: CUDADataFormats/BeamSpot/src/BeamSpotCUDA.cc (hunk 1 above)
Comment (waredjeb) [reply CUDADataFormats/BeamSpot/src/BeamSpotCUDA.cc]: So could I leave the ```cudaMemcpy``` with the direction defined?

Comment (makortel) [reply CUDADataFormats/BeamSpot/src/BeamSpotCUDA.cc]: > So could I leave the `cudaMemcpy` with the direction defined?

I'm fine with that.

**Comment by fwyzard** [comment]
:construction: Validation running at fu-c2a02-35-03:/data/user/fwyzard/patatrack/validation/run.NJwnSq9tJ7 ...

**Comment by fwyzard** [comment]
:construction: Validation running at fu-c2a02-35-01:/data/user/fwyzard/patatrack/validation/run.OKnoBYsYUS ...

File: DataFormats/Math/test/CholeskyInvert_t.cu [hunk 2]
```diff
@@ -162,8 +163,9 @@ void go(bool soa) {
 
 #ifndef DOPROF
       cudautils::launch(invertSeq<MX, DIM>, {blocksPerGrid, threadsPerBlock}, (MX *)(m_d.get()), SIZE);
+      cuda::launch(invertSeq<MX, DIM>, {blocksPerGrid, threadsPerBlock}, (MX *)(m_d.get()), SIZE);
```
Comment (fwyzard) [inline DataFormats/Math/test/CholeskyInvert_t.cu]: Looks like this line is there by mistake - possibly a leftover from the conflict resolution ?

File: HeterogeneousCore/CUDAUtilities/test/copyAsync_t.cpp [hunk 3]
```diff
@@ -21,8 +21,9 @@ TEST_CASE("copyAsync", "[cudaMemTools]") {
       auto host = cudautils::make_host_unique<int>(stream);
 
       cudautils::copyAsync(device, host_orig, stream);
-      cuda::memory::async::copy(host.get(), device.get(), sizeof(int), stream);
       cudaCheck(cudaStreamSynchronize(stream));
+      cudaMemcpyAsync(host.get(), device.get(), sizeof(int), cudaMemcpyDeviceToHost, stream);
+      stream.synchronize();
```
Comment (fwyzard) [inline HeterogeneousCore/CUDAUtilities/test/copyAsync_t.cpp]: These two lines seem incorrect:
  - `cudaMemcpyAsync` should come before `cudaStreamSynchronize`
  - `stream.synchronize();` seems spurious, possibly due to incorrect conflict resolution

File: HeterogeneousCore/CUDAUtilities/test/copyAsync_t.cpp [hunk 4]
```diff
@@ -40,8 +41,9 @@ TEST_CASE("copyAsync", "[cudaMemTools]") {
 
       SECTION("Copy all") {
         cudautils::copyAsync(device, host_orig, N, stream);
-        cuda::memory::async::copy(host.get(), device.get(), N * sizeof(int), stream);
         cudaCheck(cudaStreamSynchronize(stream));
+	cudaMemcpyAsync(host.get(), device.get(), N *  sizeof(int), cudaMemcpyDeviceToHost, stream);
```
Comment (fwyzard) [inline HeterogeneousCore/CUDAUtilities/test/copyAsync_t.cpp]: These two lines seem incorrect:
  - `cudaMemcpyAsync` should come before `cudaStreamSynchronize`
  - `stream.synchronize();` seems spurious, possibly due to incorrect conflict resolution

File: HeterogeneousCore/CUDAUtilities/test/copyAsync_t.cpp [hunk 5]
```diff
@@ -53,8 +55,9 @@ TEST_CASE("copyAsync", "[cudaMemTools]") {
 
       SECTION("Copy some") {
         cudautils::copyAsync(device, host_orig, 42, stream);
-        cuda::memory::async::copy(host.get(), device.get(), 42 * sizeof(int), stream);
         cudaCheck(cudaStreamSynchronize(stream));
+	cudaMemcpyAsync(host.get(), device.get(), 42 * sizeof(int), cudaMemcpyDeviceToHost, stream);
```
Comment (fwyzard) [inline HeterogeneousCore/CUDAUtilities/test/copyAsync_t.cpp]: These two lines seem incorrect:
  - `cudaMemcpyAsync` should come before `cudaStreamSynchronize`
  - `stream.synchronize();` seems spurious, possibly due to incorrect conflict resolution

File: RecoLocalTracker/SiPixelClusterizer/test/gpuClustering_t.h [hunk 6]
```diff
@@ -354,10 +351,10 @@ int main(void) {
     std::cout << "found " << nModules << " Modules active" << std::endl;
 
 #ifdef __CUDACC__
-    cuda::memory::copy(h_id.get(), d_id.get(), size16);
-    cuda::memory::copy(h_clus.get(), d_clus.get(), size32);
-    cuda::memory::copy(&nclus, d_clusInModule.get(), MaxNumModules * sizeof(uint32_t));
-    cuda::memory::copy(&moduleId, d_moduleId.get(), nModules * sizeof(uint32_t));
+    cudaMemcpy(h_id.get(), d_id.get(), size16, cudaMemcpyDeviceToHost);
+    cudaMemcpy(h_clus.get(), d_clus.get(), size16, cudaMemcpyDeviceToHost);
```
Comment (fwyzard) [inline RecoLocalTracker/SiPixelClusterizer/test/gpuClustering_t.h]: here `size16` was originally `size32` ?

File: RecoPixelVertexing/PixelVertexFinding/test/VertexFinder_t.h [hunk 7]
```diff
@@ -223,9 +223,9 @@ int main() {
 
 #ifdef __CUDACC__
       cudautils::launch(fitVertices, {1, 1024 - 256}, onGPU_d.get(), ws_d.get(), 50.f);
-      cuda::memory::copy(&nv, LOC_ONGPU(nvFinal), sizeof(uint32_t));
-      cuda::memory::copy(nn, LOC_ONGPU(ndof), nv * sizeof(int32_t));
-      cuda::memory::copy(chi2, LOC_ONGPU(chi2), nv * sizeof(float));
+      cudaMemcpy(&nv, LOC_ONGPU(nvFinal), sizeof(uint32_t), cudaMemcpyDeviceToHost);
+      cudaMemcpy(nn, LOC_ONGPU(ndof), nv * sizeof(uint32_t), cudaMemcpyDeviceToHost);
```
Comment (fwyzard) [inline RecoPixelVertexing/PixelVertexFinding/test/VertexFinder_t.h]: here `uint32_t` was originally `int32_t`

File: RecoLocalTracker/SiPixelClusterizer/test/gpuClustering_t.h (hunk 6 above)
Comment (waredjeb) [reply RecoLocalTracker/SiPixelClusterizer/test/gpuClustering_t.h]: I verified, only ```cudaMemcpy(h_clus.get(), d_clus.get(), size16, cudaMemcpyDeviceToHost);``` has ```size32```, fixed.

File: RecoPixelVertexing/PixelVertexFinding/test/VertexFinder_t.h (hunk 7 above)
Comment (waredjeb) [reply RecoPixelVertexing/PixelVertexFinding/test/VertexFinder_t.h]: I verified, only ```cudaMemcpy(nn, LOC_ONGPU(ndof), nv * sizeof(uint32_t), cudaMemcpyDeviceToHost);``` was originally ```int32_t```

**Comment by fwyzard** [comment]
:construction: Validation running at fu-c2a02-35-01:/data/user/fwyzard/patatrack/validation/run.GKFLRK8VN4 ...

**Comment by fwyzard** [comment]
@VinInn could you have a look at this PR ?

The changes should be only technical (moving from the `cuda::memory::..` wrappers to the standard `cudaMemcpy()` etc. functions), but we observe a non negligible change in the TTbra realistic performance for the tracks associate to the PV:


  | reference-10824.5 | development-10824.5 | development-10824.52 | testing-10824.52
-- | -- | -- | -- | --
Number of TrackingParticles (after cuts) |   | 4605 | 4950 | 5017
Number of matched TrackingParticles |   | 2346 | 2757 | 2790
Number of tracks |   | 3410 | 4371 | 4416
Number of true tracks |   | 3025 | 3860 | 3905
Number of fake tracks |   | 385 | 511 | 511
Number of pileup tracks |   | 0 | 0 | 0
Number of duplicate tracks |   | 44 | 0 | 0

while there doesnt seem to be any change in the overall tracks:

  | reference-10824.5 | development-10824.5 | development-10824.52 | testing-10824.52
-- | -- | -- | -- | --
Efficiency | 0.5128 | 0.5252 | 0.5818 | 0.5818
Number of TrackingParticles (after cuts) | 5530 | 5320 | 5320 | 5320
Number of matched TrackingParticles | 2836 | 2794 | 3095 | 3095
Fake rate | 0.0472 | 0.0479 | 0.0212 | 0.0212
Duplicate rate | 0.0150 | 0.0152 | 0.0003 | 0.0003
Number of tracks | 32648 | 32656 | 39763 | 39763
Number of true tracks | 31108 | 31093 | 38921 | 38920
Number of fake tracks | 1540 | 1563 | 842 | 843
Number of pileup tracks | 27279 | 27270 | 34468 | 34467
Number of duplicate tracks | 491 | 495 | 12 | 12

File: HeterogeneousCore/CUDAUtilities/interface/copyAsync.h [hunk 8]
```diff
@@ -17,15 +18,15 @@ namespace cudautils {
...
+    cudaCheck(cudaMemcpyAsync(dst.get(), src.get(), sizeof(T), cudaMemcpyHostToDevice, stream));
   }
 
   template <typename T>
   inline void copyAsync(cudautils::host::unique_ptr<T>& dst,
                         const cudautils::device::unique_ptr<T>& src,
                         cudaStream_t stream) {
     static_assert(std::is_array<T>::value == false, "For array types, use the other overload with the size parameter");
-    cuda::memory::async::copy(dst.get(), src.get(), sizeof(T), stream);
+    cudaCheck(cudaMemcpyAsync(dst.get(), src.get(), sizeof(T), cudaMemcpyHostToDevice, stream));
```
Comment (VinInn) [inline HeterogeneousCore/CUDAUtilities/interface/copyAsync.h]: this is device2host

File: HeterogeneousCore/CUDAUtilities/interface/copyAsync.h [hunk 9]
```diff
@@ -34,15 +35,15 @@ namespace cudautils {
...
+    cudaCheck(cudaMemcpyAsync(dst.get(), src.get(), nelements * sizeof(T), cudaMemcpyHostToDevice, stream));
   }
 
   template <typename T>
   inline void copyAsync(cudautils::host::unique_ptr<T[]>& dst,
                         const cudautils::device::unique_ptr<T[]>& src,
                         size_t nelements,
                         cudaStream_t stream) {
-    cuda::memory::async::copy(dst.get(), src.get(), nelements * sizeof(T), stream);
+    cudaCheck(cudaMemcpyAsync(dst.get(), src.get(), nelements * sizeof(T), cudaMemcpyHostToDevice, stream));
```
Comment (VinInn) [inline HeterogeneousCore/CUDAUtilities/interface/copyAsync.h]: ditto

File: HeterogeneousCore/CUDAUtilities/interface/copyAsync.h (hunk 8 above)
Comment (makortel) [reply HeterogeneousCore/CUDAUtilities/interface/copyAsync.h]: And "Calling `cudaMemcpy()` with dst and src pointers that do not match the direction of the copy results in an undefined behavior." (*), so specifying the direction explicitly is actually harmful?

(*) https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1gc263dbe6574220cc776b45438fc351e8

Comment (VinInn) [reply HeterogeneousCore/CUDAUtilities/interface/copyAsync.h]: Indeed. I think we agreed to remove all explicit directions.

Comment (fwyzard) [reply HeterogeneousCore/CUDAUtilities/interface/copyAsync.h]: > Calling `cudaMemcpyAsync()` with `dst` and `src` pointers that do not match the direction of the copy results in an undefined behavior.

I thought it was supposed to crash...

**Comment by fwyzard** [comment]
:construction: Validation running at fu-c2a02-35-01:/data/user/fwyzard/patatrack/validation/run.SglAPpu5Wz ...

**Comment by fwyzard** [comment]
OK, now it looks better.
The same summary comparison now give identical results:

  | reference-10824.5 | development-10824.5 | development-10824.52 | testing-10824.52
-- | -- | -- | -- | --
Number of TrackingParticles (after cuts) |   | 4605 | 5017 | 5017
Number of matched TrackingParticles |   | 2346 | 2790 | 2790
Number of tracks |   | 3410 | 4416 | 4416
Number of true tracks |   | 3025 | 3905 | 3905
Number of fake tracks |   | 385 | 511 | 511
Number of pileup tracks |   | 0 | 0 | 0
Number of duplicate tracks |   | 44 | 0 | 0

and all the others show identical or almost identical results.

**Pull request merged**